File: android-sms-reader/app/src/main/java/com/rjworldbd/smsreader/receiver/SmsReceiver.java
package com.rjworldbd.smsreader.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;
import com.rjworldbd.smsreader.data.PaymentDbHelper;
import com.rjworldbd.smsreader.parser.PaymentSmsParser;
import com.rjworldbd.smsreader.parser.PaymentSmsParser.ParsedPayment;
import com.rjworldbd.smsreader.service.SmsReaderForegroundService;
import com.rjworldbd.smsreader.sync.SyncManager;

/**
 * BroadcastReceiver for incoming SMS.
 * STRICT ENFORCEMENT:
 * 1. Only processes bKash, Nagad, Rocket, Upay payment SMS.
 * 2. All other messages (personal, promo, OTP) are immediately ignored.
 * 3. Never stores raw SMS text anywhere.
 * 4. Checks duplicate TrxID before insertion.
 * 5. Updates persistent foreground notification counts.
 */
public class SmsReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent == null || intent.getAction() == null) return;
        if (!"android.provider.Telephony.SMS_RECEIVED".equals(intent.getAction())) return;

        Bundle bundle = intent.getExtras();
        if (bundle == null) return;

        Object[] pdus = (Object[]) bundle.get("pdus");
        if (pdus == null || pdus.length == 0) return;

        String format = bundle.getString("format");
        StringBuilder bodyBuilder = new StringBuilder();
        String sender = null;
        long timestamp = System.currentTimeMillis();

        for (Object pdu : pdus) {
            SmsMessage message;
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
                message = SmsMessage.createFromPdu((byte[]) pdu, format);
            } else {
                message = SmsMessage.createFromPdu((byte[]) pdu);
            }
            if (message != null) {
                if (sender == null) {
                    sender = message.getOriginatingAddress();
                    timestamp = message.getTimestampMillis();
                }
                bodyBuilder.append(message.getMessageBody());
            }
        }

        if (sender == null || bodyBuilder.length() == 0) return;

        // Parse and detect MFS payment SMS
        ParsedPayment payment = PaymentSmsParser.parse(sender, bodyBuilder.toString(), timestamp);
        if (payment == null) {
            // Non-payment SMS or unsupported provider: discard immediately
            return;
        }

        // Check for duplicate TrxID and insert
        final Context appContext = context.getApplicationContext();
        PaymentDbHelper db = PaymentDbHelper.getInstance(appContext);
        db.insertParsedPayment(payment);

        // Update foreground service counts
        int pending = db.getPendingCount();
        int synced = db.getSyncedCount();
        SmsReaderForegroundService.updateCounts(appContext, pending, synced);

        // Immediately sync this detected payment to Firebase Realtime Database with zero delay
        SyncManager.syncPaymentImmediately(appContext, payment);
        // Also trigger queue sync for any pending offline records
        SyncManager.triggerSync(appContext);
    }
}

File: android-sms-reader/app/src/main/java/com/rjworldbd/smsreader/receiver/SmsReceiver.kt
package com.rjworldbd.smsreader.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.provider.Telephony
import androidx.work.*
import com.rjworldbd.smsreader.data.local.PaymentDatabase
import com.rjworldbd.smsreader.data.model.PaymentVerificationEntity
import com.rjworldbd.smsreader.parser.PaymentSmsParser
import com.rjworldbd.smsreader.sync.PaymentSyncWorker
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * BroadcastReceiver for incoming SMS.
 * STRICT RULES:
 * 1. Only processes bKash, Nagad, Rocket, Upay payment SMS.
 * 2. All other SMS are ignored immediately.
 * 3. Never stores or uploads full SMS body.
 * 4. Checks duplicate TrxID to prevent double submission.
 * 5. Queues into local Room database and triggers WorkManager sync.
 */
class SmsReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != Telephony.Sms.Intents.SMS_RECEIVED_ACTION) return

        val messages = Telephony.Sms.Intents.getMessagesFromIntent(intent)
        if (messages.isNullOrEmpty()) return

        // Extract sender and combined body
        val sender = messages[0].originatingAddress ?: return
        val bodyBuilder = StringBuilder()
        var timestamp = System.currentTimeMillis()

        for (msg in messages) {
            bodyBuilder.append(msg.messageBody)
            timestamp = msg.timestampMillis
        }
        val fullBody = bodyBuilder.toString()

        // Step 1: Parse and detect strictly supported payment SMS
        val parsed = PaymentSmsParser.parse(sender, fullBody, timestamp) ?: return // Ignore non-payment SMS

        // Step 2: Coroutine for local database operations
        val pendingResult = goAsync()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val db = PaymentDatabase.getInstance(context)
                val dao = db.paymentDao()

                // Duplicate Check by Transaction ID across ALL stored transactions
                val existing = dao.findByTransactionId(parsed.transactionId)
                if (existing != null) {
                    if (existing.syncStatus == "pending_sync") {
                        val constraints = Constraints.Builder()
                            .setRequiredNetworkType(NetworkType.CONNECTED)
                            .build()
                        val syncRequest = OneTimeWorkRequestBuilder<PaymentSyncWorker>()
                            .setConstraints(constraints)
                            .build()
                        WorkManager.getInstance(context).enqueue(syncRequest)
                    }
                    return@launch
                }

                val paymentId = "PAY-${System.currentTimeMillis().toString(36).uppercase()}"
                val invoiceId = "INV-${(System.currentTimeMillis() % 1000000)}"

                val entity = PaymentVerificationEntity(
                    paymentId = paymentId,
                    invoiceId = invoiceId,
                    userId = "android_sms_detected",
                    userType = "customer",
                    paymentMethod = parsed.paymentMethod,
                    expectedAmount = parsed.amount,
                    transactionId = parsed.transactionId,
                    status = "pending",
                    senderNumber = parsed.senderNumber,
                    receivedAmount = parsed.amount,
                    verifiedAt = null,
                    createdAt = parsed.timestamp,
                    rejectionReason = null,
                    syncStatus = "pending_sync",
                    isOfflineCreated = false
                )

                // Save to local queue
                dao.insertOrUpdate(entity)

                // Update Persistent Notification with current pending and synced count
                val pendingCount = dao.getPendingCount()
                val syncedCount = dao.getSyncedCount()
                com.rjworldbd.smsreader.service.SmsReaderForegroundService.updateCounts(context, pendingCount, syncedCount)

                // Trigger network sync worker (WorkManager with network constraint)
                val constraints = Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build()

                val syncRequest = OneTimeWorkRequestBuilder<PaymentSyncWorker>()
                    .setConstraints(constraints)
                    .build()

                WorkManager.getInstance(context).enqueue(syncRequest)
            } finally {
                pendingResult.finish()
            }
        }
    }
}
